Introduce "reporting endpoint" to COEP

yutakahirano opened the pull request:
This change introduces "reporting endpoint" concept which will be used as an
reporting endpoint for potential COEP violations.

Reporting behavior will be specified in future changes.

Discussed at https://github.com/whatwg/html/issues/5100.

yutakahirano committed:
Introduce "reporting endpoint" to COEP

This change introduces "reporting endpoint" concept which will be used as an
reporting endpoint for potential COEP violations.

Reporting behavior will be specified in future changes.

Discussed at https://github.com/whatwg/html/issues/5100.

yutakahirano committed:
fix

zcorpan commented:
This should probably follow the convention in https://github.com/mikewest/corpp/pull/2

yutakahirano commented:
I'll do that after that change lands.

mikewest commented on index.bs:
````diff
@@ -148,6 +148,23 @@ this header if it contains an invalid value. Likewise, user agents MUST ignore t
+[[!I-D.ietf-httpbis-header-structure]] Its ABNF is:
+
+```
+Cross-Origin-Embedder-Policy-Report-To = sh-token
````
Why would we do this as a separate header rather than a parameter on the existing header? That is, `Cross-Origin-Embedder-Policy: require-corp; report-to=group` is simple, consistent with CSP, and (IMO) aesthetically better than creating a new header.

@annevk, as I think you prefer the way it's defined here. :)

yutakahirano replied:
Done.

mikewest commented on index.bs:
```diff
@@ -258,7 +297,29 @@ so, HTML is patched as follows:
         ::  Return <var ignore>worker global scope</var>'s [=WorkerGlobalScope/embedder policy=].
 
-3.  The [$create a new browsing context$] algorithm sets the [=document/embedder policy=] for
+3. An <dfn>embeedder policy reporting endpoint</dfn> is a nullable string.
```
```suggestion
3. An <dfn>embedder policy reporting endpoint</dfn> is a nullable string.
```

yutakahirano replied:
Done.

mikewest commented on index.bs:
```diff
@@ -258,7 +297,29 @@ so, HTML is patched as follows:
-3.  The [$create a new browsing context$] algorithm sets the [=document/embedder policy=] for
+3. An <dfn>embeedder policy reporting endpoint</dfn> is a nullable string.
+
+4. The [=/embedder policy reporting endpoint=] is persisted on a number of objects:
```
Similar to the above, it seems more reasonable to me to make the embedder policy a struct of the value and the reporting endpoint.

yutakahirano replied:
Done.

mikewest reviewed (commented):
Thanks for this patch! I have a few questions:

1.  Do we need a report-only mode? ( @arturjanc )
2.  We could do this in one header rather than two (or four, if `report-only` is a thing). Wouldn't a param be simpler?

yutakahirano commented:
Thank you!

> 1. Do we need a report-only mode? ( @arturjanc )

What do you mean by report-only mode here? I'm planning to introduce the following behaviors:

 1. If COEP is "none" and the reporting endpoint is null: do nothing
 1. If COEP is "require-corp" and the reporting endpoint is null: COEP is effective without reporting
 1. If COEP is "none" and the reporting endpoint is set: Do not block things but report "would-be" violations to the server.
 1. If COEP is "require-corp" and the reporting endpoint is set: COEP is effective, and report violations to the server.

I thought the third settings would be the "report-only" mode.

> 2. We could do this in one header rather than two (or four, if report-only is a thing). Wouldn't a param be simpler?

I'm fine with this.

mikewest commented:
> I thought the third settings would be the "report-only" mode.

Got it. I'm fine with that. It's a little distinct from CSP, but that's not necessarily a bad thing.

If we go that route, the single-header variant I've suggested above would require introducing a new token. Perhaps `none`, or something more explicit (like `report-corp-violations`)?

yutakahirano committed:
fix

yutakahirano committed:
fix

yutakahirano commented:
I fixed the header parsing part (I haven't changed the properties on context objects yet). Mike, can you take a look at that part?

yutakahirano committed:
fix

yutakahirano committed:
fix

yutakahirano commented:
Now I think I addressed your comment. PTAL.

mikewest commented on index.bs:
````diff
@@ -138,10 +138,11 @@ server to declare an embedder policy for a given document. It is a [=Structured
 
 ```
-Cross-Origin-Embedder-Policy = sh-token
+Cross-Origin-Embedder-Policy = sh-token [parameter]
````
It's probably simpler to just use the item syntax:

```suggestion
Cross-Origin-Embedder-Policy = sh-item
```

mikewest commented on index.bs:
```diff
@@ -138,10 +138,11 @@ server to declare an embedder policy for a given document. It is a [=Structured
 
-The only currently valid `Cross-Origin-Embedder-Policy` value is "`require-corp`".
+The `Cross-Origin-Embedder-Policy` value consists of one token (either "`require-corp`" or
+"`unsafe-none`") with possibly a parameter which specify an endpoint for violation reporting.
```
Nit:

```suggestion
"`unsafe-none`") which may have a parameter specifying an endpoint for violation reporting.
```

mikewest commented on index.bs:
```diff
@@ -155,28 +156,38 @@ Parsing {#parsing}
-1.  Let |policy| be "`unsafe-none`".
-
-2.  Let |header| be the result of [=header list/getting=] `Cross-Origin-Embedder-Policy` from
+1.  Let |header| be the result of [=header list/getting=] `Cross-Origin-Embedder-Policy` from
```
You can change step 1-4 to use "[get a structured header](https://fetch.spec.whatwg.org/#concept-header-list-get-structured-header)", passing in `Cross-Origin-Embedder-Policy` and `item`.

That will give you |parsed item| directly (which, according to https://tools.ietf.org/html/draft-ietf-httpbis-header-structure-15#section-4.2.3, is a `(bare_item, parameters)` tuple.

zcorpan replied:
It seems to me we could avoid duplicated effort of figuring out how to correctly get structured headers by getting https://github.com/mikewest/corpp/pull/2 done and then doing the same thing here.

Are we ready to introduce SH parameters now? @annevk @mnot

mikewest replied:
Parameters are part of the SH spec, which @mnot asserts elsewhere is ready for last call. I think it's reasonable to expect that they'll be part of headers going forward.

yutakahirano replied:
Done.

mikewest commented on index.bs:
```diff
@@ -155,28 +156,38 @@ Parsing {#parsing}
+5.  Let |value| be the first component of |parsed item|.
 
-    1.  Set |header| to the result of [=isomorphic decoding=] |header|.
+6.  Let |parameters| be the second component of |parsed item|.
```
`value` and `parameters` are `|parsed item|'s bare_item` and `|parsed item|'s parameters` respectively.

yutakahirano replied:
Done.

mikewest commented on index.bs:
```diff
@@ -155,28 +156,38 @@ Parsing {#parsing}
-        ISSUE(httpwg/http-extensions#662): The ASCII requirements of Structured Headers are
-        somewhat unclear to me. Do we need to check whether |header| is an [=ASCII string=] before
-        passing it into the parsing algorithm?
+9.  For each |name| with a value of |value| in |parameters:
```
SH defines parameters as a map, so you can just grab the specific item you want. That is, something like the following should work instead of this loop.

>  If |parsed item|'s parameters["report-to"] [=map/exists=], set |endpoint| to |parsed item|'s parameters["report-to"].

yutakahirano replied:
Done.

mikewest commented on index.bs:
```diff
@@ -238,16 +249,20 @@ restrictions it asserts. Likewise, user agents MUST also take into account the e
+    1.  A string (<dfn for="embedder policy">value</dfn>) with one of the following values:
+        "`unsafe-none`", "`require-corp`", initially "`unsafe-none`".
+
+    1.  A nullable string (<dfn for="embedder plicy">reporting endpoint</dfn>), initially null.
```
```suggestion
    1.  A nullable string (<dfn for="embedder policy">reporting endpoint</dfn>), initially null.
```

mikewest replied:
Elsewhere we describe this as "A string or `null`". I'm not sure "nullable string" is a concept HTML knows about (it doesn't appear in a quick search). @annevk might have opinions about whether it should. :)

annevk replied:
I'd prefer sticking with a string or null.

yutakahirano replied:
Done.

mikewest reviewed (commented):
Looks pretty reasonable to me. I'd like @annevk's opinion as well, as I think he disagrees about the header syntax, and we should probably chat about it rather than just landing this. :)

arturjanc commented:
> 1. Do we need a report-only mode? ( @arturjanc )

I think you've settled this already, but just mentioning that the approach above looks good to me (the 4 behaviors mentioned by @yutakahirano, possibly updated if we decide to go with the single-header variant).

In general, I'd posit that there are a lot of benefits to having a report-only mode for everything :)

annevk commented:
Does anyone have some examples to evaluate? I find the diff a little hard to read.

yutakahirano committed:
Update index.bs

Accept mikewset's suggestion

Co-Authored-By: Mike West <mike@mikewest.org>

yutakahirano committed:
Update index.bs

Accept mikewest's suggestion

Co-Authored-By: Mike West <mike@mikewest.org>

yutakahirano committed:
fix

yutakahirano committed:
fix

yutakahirano commented:
This change:

 - introduces "unsafe-none" as a value of the COEP header.
 - introduces "report-to" parameter to the COEP header
 - introduces the _embedder policy_ concept which consists of COEP value and the reporting endpoint.

COEP headers now look like:

 - `unsafe-none` : Do nothing.
 - `unsafe-none; report-to=endpoint` Do not block things, but report would-be violations using "endpoint" reporting endpoint.
 - `require-corp`: COEP is effective without reporting
 - `require-corp; report-to=endpoint`: COEP is effective and report violations using "endpoint" reporting endpoint.

This change does __NOT__ include how the reporting endpoint is used. I'm going to publish another PR for that part.

@annevk, does this summary help?

yutakahirano commented:
I would like the syntax to be aligned with COOP reporting. @camillelamy, are you fine with the proposal?

annevk commented:
The summary is great, thanks again!

After discussion on IRC I think I'm okay with this approach, as we'll likely keep COOP and COEP scoped to a simple enumerated value. If anyone foresees them using more complex values, now would be the time to bring that up. (If the value can be more complex, I think you want more headers eventually, as explained in https://github.com/w3c/webappsec-feature-policy/issues/362#issuecomment-582302307.)

zcorpan commented:
@yutakahirano the `:` is supposed to be `;` , correct?

yutakahirano commented:
@zcorpan Yes, thanks. Fixed.

zcorpan commented on index.bs:
```diff
@@ -138,10 +138,11 @@ server to declare an embedder policy for a given document. It is a [=Structured
 
-The only currently valid `Cross-Origin-Embedder-Policy` value is "`require-corp`".
+The `Cross-Origin-Embedder-Policy` value consists of one token (either "`require-corp`" or
+"`unsafe-none`") which may have a parameter specifying an endpoint for violation reporting.
```
Should the value type of the parameter be specified?

https://tools.ietf.org/html/draft-ietf-httpbis-header-structure-14#section-2 suggests yes, but the type is not passed along the parsing algorithm, as far as I can tell. @mnot how is this intended to work?

mikewest replied:
It does seem like we should specify the type. Perhaps "specifying a [=structured-header/string=] which represents the endpoint for ..."?

For parsing, I think that the structured headers spec parses parameter values as `sh-item`. I suppose we're on our own to validate them.

yutakahirano replied:
Done. I also added one more check to the parsing logic.

mikewest reviewed (approved):
I'm happy with this % the discussion of parameter types. If @camillelamy and team are happy too, I'll pull it in.

yutakahirano committed:
fix

yutakahirano committed:
fix

camillelamy commented:
Our proposal for COOP does not have the same syntax. Currently we're looking at 3 headers:

- Cross-Origin-Opener-Policy: the policy to enforce
- Cross-Origin-Opener-Policy-Report-Only: allows to specify a policy which is not enforced and for which we only provide reports
- Cross-Origin-Opener-Policy-Report-Endpoint: where to send the reports (which would be generated for both the COOP enforced and the COOP report-only mode).

We could move to a model closer to the one you propose. but we'd still have two headers:

- Cross-Origin-Opener-Policy : value; report-to=endpoint
- Cross-Origin-Opener-Policy-Report-Only : value; report-to=endpoint

This is because COOP has 3 values, so we cannot make the assumption that unsafe-none + an endpoint means we want a report only mode for same-origin (it could be same-origin-allow-popups).

The problem I see with the current approach for COEP is that if we want to add another value to the header, the report-only mode doesn't work anymore (like it wouldn't with COOP). If we foresee the need to eventually add another value, we should allow the developer to specify explicitly which policy they want report-only mode for instead of making this deduction.

annevk commented:
Thanks Camille! @yutakahirano had an idea for that problem when COEP gets to it, but now you contrast it with COOP I think we ought to solve it now so we can have a consistent reporting story across these two features.

yutakahirano commented:
Thank you, Camille.

Does the following work?

cross-origin-opener-policy: _policy_; reporting_as=_another policy_; reporting_to=_endpoint_

I'm also fine with the two-headers version. The three-headers version seems too verbose to me.

camillelamy commented:
I think the two headers version matches what CSP is doing, right? If I'm not mistaken, CSP has two headers: Content-Security-Policy and Content-Security-Policy-Report-Only. It might be easier for developers to have the same model for COOP and COEP?

yutakahirano committed:
fix

yutakahirano commented:
Fixed. Now we have two headers: "Cross-Origin-Embedder-Policy" (COEP) and "Cross-Origin-Embedder-Policy-Report-Only" (COEP-Report-Only). I removed "unsafe-none" value from the header(s).

Cross-Origin-Embedder-Policy-Report-Only is ignored if there is (valid) Cross-Origin-Embedder-Policy attached; Hence there are four modes:

- `COEP: require-corp`: CORP-Report-Only is ignored. COEP is effective. Violations are not reported.
- `COEP: require-corp; report-to="endpoint"`: CORP-Report-Only is ignored. COEP is effective. Violations are reported to "endpoint".
- `(COEP: missing,) COEP-Report-Only: require-corp`: No effect
- `(COEP: missing,) COEP-Report-Only: require-corp; report-to="endpoint"`: Do not block things. Potential violations are reported to "endpoint".
- `(COEP: missing, COEP-Report-Only: missing)`: No effect

The "embedder policy" concept consists of value (either "require-corp" and "unsafe-none"), and reporting endpoint. Unlike the pair of headers above, this concept covers all modes by itself.

- `{value: "require-corp", reporting endpoint: null}`: COEP is effective. Violations are not reported.
- `{value: "require-corp", reporting endpoint: "endpoint"}`: COEP is effective. Violations are reported to "endpoint".
- `{value: "unsafe-none", reporting endpoint: null}`: No effect
- `{value: "unsafe-none", reporting endpoint: "endpoint"}`: Do not block things. Potential violations are reported to "endpoint".

The embedder policy concept is not exposed to HTTP, so if the syntax grows we can change the concept without breaking things.

@mikewest @annevk @camillelamy Are you fine with the latest change?

camillelamy commented:
Thanks! I am happy with the latest version.

yutakahirano commented:
@annevk @mikewest can you take a look at the latest change? Thanks!

mikewest commented on index.bs:
```diff
@@ -155,28 +179,31 @@ Parsing {#parsing}
+        |endpoint| to |parsed item|'s parameters["report-to"].
 
-    1.  Set |header| to the result of [=isomorphic decoding=] |header|.
+    2.  Return a new [=/embedder policy=] whose [=embedder policy/value] is |parsed item|'s
```
This means that `COEPRO` is ignored in the presence of `COEP`. This is probably fine for the moment, since there's only one value, but would make it difficult to support enforcement of one value and reporting on another if we have multiple values in the future.

Perhaps add a note to that effect? I'm not sure it's worth adding complexity until we need it.

yutakahirano replied:
I'd say COEPRO is ignored if its value is equal to or weaker than the COEP value. That rule is natural, and we'll want to keep the rule even when we introduce more values. What do you think?

mikewest replied:
1. That's not the way CSP works.
2. What do we do with `COEP: require-corp\nCOEPRO: require-corp; report-to=whatever`? Should that not send reports?

yutakahirano replied:
Hmm... so you prefer having two values and endpoints. I'll try to modify the logic.

mikewest replied:
I think we will need to have two values and endpoints to support both enforce and report-only at the same time. I don't think that's actually relevant to any developer's life until we have multiple values (I have `require-cors` in the back of my head (because a value with a one-character difference to create radical behavioral differences is a great spelling option)), so I wouldn't mind explicitly punting on it. But doing it now certainly wouldn't be wasted effort.

yutakahirano replied:
Fixed. Now the embedder policy concept has four members.
 - value
 - reporting endpoint
 - report only value
 - report only reporting endpoint

yutakahirano replied:
reporting endpoint / report only reporting endpoint is not set if value / report only value is not "require-corp" but it doesn't affect the behavior because we'll have nothing to report for "unsafe-none".

mikewest reviewed (approved):
LGTM % one comment.

yutakahirano committed:
fix

yutakahirano committed:
fix

yutakahirano commented:
ping?

mikewest reviewed (approved):
LGTM. Let's land this and continue iterating. :)

yutakahirano commented:
Thank you!
